### PR TITLE
[in_app_purchase] Add support for SKPaymentQueueDelegate and showPriceConsentIfNeeded

### PR DESCRIPTION
**Pull request by mvanbeusekom**
Enables developers to register a custom implementation of the `SKPaymentQueueDelegateWrapper` class and control how changes regarding subscriptions should be processed. 

Fixes flutter/flutter#85129

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I [updated pubspec.yaml](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates) with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code

### CODE REVIEW

File: packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAPPaymentQueueDelegate.m [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+                                if (result && [result isKindOfClass:[NSNumber class]]) {
+                                  shouldContinue = [(NSNumber *)result boolValue];
+                                }
+
+                                dispatch_semaphore_signal(semaphore);
+                              }];
+
+  // The client should respond within 1 second otherwise continue
+  // with default value.
+  dispatch_semaphore_wait(semaphore, dispatch_time(DISPATCH_TIME_NOW, NSEC_PER_SEC));
```
Comment (renefloor) [inline packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAPPaymentQueueDelegate.m:48]: This means the Flutter side should not first show a dialog to the user saying 'do you want to get the price consent window'. The iOS native implementation is sync, but the Flutter implementation is async. The async-ness might give devs the idea that they can wait for a user interaction.

File: packages/in_app_purchase/in_app_purchase_ios/lib/src/store_kit_wrappers/sk_payment_queue_delegate_wrapper.dart [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+      Future.value(true);
+
+  /// Called by the system to check whether to immediately show the price
+  /// consent form.
+  ///
+  /// The default return value is `true`. This will inform the system to display
+  /// the price consent sheet when the subscription price has been changed in
+  /// App Store Connect and the subscriber has not yet taken action. See the
+  /// documentation in StoreKit's [`[-SKPaymentQueueDelegate shouldShowPriceConsent:]`](https://developer.apple.com/documentation/storekit/skpaymentqueuedelegate/3521328-paymentqueueshouldshowpriceconse?language=objc).
+  Future<bool> shouldShowPriceConsent() => Future.value(true);
```
Comment (renefloor) [inline packages/in_app_purchase/in_app_purchase_ios/lib/src/store_kit_wrappers/sk_payment_queue_delegate_wrapper.dart:34]: See my other comment, why wouldn't you make this function sync. The iOS side expects a result within a second, so you can't do user interactions or api calls.
```suggestion
  bool shouldShowPriceConsent() => true;
```

**Review by renefloor — CHANGES_REQUESTED** [review]
Wouldn't it be better to make the dart signatures sync?

File: packages/in_app_purchase/in_app_purchase_ios/ios/Classes/InAppPurchasePlugin.m [hunk 3]
```diff
@@ -103,6 +111,14 @@ - (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result
...
   } else if ([@"-[SKPaymentQueue stopObservingTransactionQueue]" isEqualToString:call.method]) {
     [_paymentQueueHandler stopObservingPaymentQueue];
+  } else if ([@"-[SKPaymentQueue registerDelegate]" isEqualToString:call.method]) {
+    [self registerPaymentQueueDelegate];
+  } else if ([@"-[SKPaymentQueue removeDelegate]" isEqualToString:call.method]) {
+    [self removePaymentQueueDelegate];
+  } else if ([@"-[SKPaymentQueue showPriceConsentIfNeeded]" isEqualToString:call.method]) {
+    if (@available(iOS 13.4, *)) {
+      [_paymentQueueHandler showPriceConsentIfNeeded];
+    }
```
Comment (renefloor) [inline packages/in_app_purchase/in_app_purchase_ios/ios/Classes/InAppPurchasePlugin.m:121]: Is it me or is the result callback missing.
`showPriceConsentIfNeeded` is a void so I would expect this:
```suggestion
    if (@available(iOS 13.4, *)) {
      [_paymentQueueHandler showPriceConsentIfNeeded];
    }
    result(nil);
```
or
```suggestion
    if (@available(iOS 13.4, *)) {
      [_paymentQueueHandler showPriceConsentIfNeeded];
      result(nil);
    }
    result(FlutterMethodNotImplemented);
```

File: packages/in_app_purchase/in_app_purchase_ios/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart [hunk 4]
```diff
@@ -70,18 +74,39 @@ class SKPaymentQueueWrapper {
...
+      await channel.invokeMethod<void>('-[SKPaymentQueue removeDelegate]');
+      paymentQueueDelegateChannel.setMethodCallHandler(null);
+    } else {
+      await channel.invokeMethod<void>('-[SKPaymentQueue registerDelegate]');
+      paymentQueueDelegateChannel
+          .setMethodCallHandler(handlePaymentQueueDelegateCallbacks);
+    }
+
+    _paymentQueueDelegate = delegate;
+  }
```
Comment (renefloor) [inline packages/in_app_purchase/in_app_purchase_ios/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart:109]: What if you register 2 delegates?
Shouldn't this contain a check `if(_paymentQueueDelegate != null)` and throw an exception "Already a delegate registered".

**Review by renefloor — CHANGES_REQUESTED** [review]

File: packages/in_app_purchase/in_app_purchase_ios/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart (hunk 4 above)
Comment (mvanbeusekom) [reply packages/in_app_purchase/in_app_purchase_ios/lib/src/store_kit_wrappers/sk_payment_queue_wrapper.dart:109]: If you register two delegates the existing one is replaced with the new one. There can always only be one delegate active. 

If you would add the check you would never be able to replace it if you wanted (although it is not likely that this is necessary).

File: packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAObjectTranslator.h [hunk 5]
```diff
@@ -29,6 +29,13 @@ NS_ASSUME_NONNULL_BEGIN
 
 + (NSDictionary *)getMapFromNSError:(NSError *)error;
 
++ (NSDictionary *)getMapFromSKStorefront:(SKStorefront *)storefront
+    API_AVAILABLE(ios(13), macos(10.15), watchos(6.2));
+
++ (NSDictionary *)getMapFromSKStorefront:(SKStorefront *)storefront
+                 andSKPaymentTransaction:(SKPaymentTransaction *)transaction
+    API_AVAILABLE(ios(13), macos(10.15), watchos(6.2));
+
```
[the remark this answers is not in the document]
Comment (mvanbeusekom) [reply packages/in_app_purchase/in_app_purchase_ios/ios/Classes/FIAObjectTranslator.h]: Yes you are right, fixed.

File: packages/in_app_purchase/in_app_purchase_ios/ios/Classes/InAppPurchasePlugin.m (hunk 3 above)
Comment (mvanbeusekom) [reply packages/in_app_purchase/in_app_purchase_ios/ios/Classes/InAppPurchasePlugin.m]: Fixed

**Review by renefloor — APPROVED** [review]
lgtm

**Pull request merged**